userspace/libsinsp/filter_evttype_resolver.h

```diff
@@ -0,0 +1,232 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <filter/parser.h>
+#include <string>
+#include <set>
+#include <memory>
+#include <functional>
+#include <stdexcept>
+
+
+/* Using a custom / optimized event type for performance optimization reasons.
+ *
+ * We found that regular set operations, even with unordered_set, slowed down Falco's 
+ * initialization by orders of magnitude, because of exhaustive evttypes scanning 
+ * when loading rulesets.
+ * For this reason the custom event type optimizes operations over sets of numbers.
+*/
+class sinsp_event_types
+{
+private:
+	using vec_t = std::vector<uint8_t>;
+	vec_t m_types{};
+
+	static inline void check_range(uint16_t e)
+	{
+	    	static const auto enum_max = get_ppm_event_max();
+		if(e > enum_max)
+		{
+			throw std::range_error("invalid event type");
+		}
+	}
+
+public:
+	sinsp_event_types(sinsp_event_types&&) = default;
+	sinsp_event_types(const sinsp_event_types&) = default;
+	sinsp_event_types& operator=(sinsp_event_types&&) = default;
+	sinsp_event_types& operator=(const sinsp_event_types&) = default;
+
+	static size_t get_ppm_event_max();
+
+	inline sinsp_event_types():
+		m_types(get_ppm_event_max() + 1, 0)
+	{
+	}
+
+	inline void insert(uint16_t e)
+	{
+		check_range(e);
+		m_types[e] = 1;
+	}
+
+	void merge(const sinsp_event_types& other)
+	{
+		for(size_t i = 0; i <= get_ppm_event_max(); ++i)
+		{
+			m_types[i] |= other.m_types[i];
+		}
+	}
+
+	void merge(const std::set<uint16_t>& other)
+	{
+		for(const auto& e : other)
+		{
+			insert(e);
+		}
+	}
+
+	inline bool contains(uint16_t e) const
+	{
+		check_range(e);
+		return m_types[e] != 0;
+	}
+
+	void clear()
+	{
+		for(auto& v : m_types)
+		{
+			v = 0;
+		}
+	}
+
+	bool equals(const sinsp_event_types& other) const
+	{
+		return m_types == other.m_types;
+	}
+
+	sinsp_event_types diff(const sinsp_event_types& other) const
+	{
+		sinsp_event_types ret;
+		for(size_t i = 0; i <= get_ppm_event_max(); ++i)
+		{
+			if(m_types[i] == 1 && other.m_types[i] == 0)
+			{
+				ret.m_types[i] = 1;
+			}
+		}
+		return ret;
+	}
+
+	sinsp_event_types intersect(const sinsp_event_types& other) const
+	{
+		sinsp_event_types ret;
+		for(size_t i = 0; i <= get_ppm_event_max(); ++i)
+		{
+			if(m_types[i] == 1 && other.m_types[i] == 1)
+			{
+				ret.m_types[i] = 1;
+			}
+		}
+		return ret;
+	}
+
+	void for_each(std::function<bool(uint16_t)> consumer) const
+	{
+		for(uint16_t i = 0; i < m_types.size(); ++i)
+		{
+			if(m_types[i] != 0)
+			{
+				if(!consumer(i))
+				{
+					return;
+				}
+			}
+		}
+	}
+};
+
+inline bool operator==(const sinsp_event_types& lhs, const sinsp_event_types& rhs)
+{
+	return lhs.equals(rhs);
+}
+
+inline bool operator!=(const sinsp_event_types& lhs, const sinsp_event_types& rhs)
+{
+	return !(lhs == rhs);
+}
+
+/*!
+	\brief Helper class for finding event types
+*/
+class filter_evttype_resolver
```

Comment:
Since the rest of the new-ish filter-related classes have their own namespace, I'd suggest making this become `libsinsp::filter::evttype_resolver`. Event better would be to make all the `evttypes` methods become simple overloaded functions under `libsinsp::filter::evttype()`, and then have only the visitor class defined as static code in the .cpp file where the functions are implemented. If you're in doubt, we can postpone this to a subsequent change along with the templatization of `sinsp_event_types`, but I think this would be a good cleanup!